File: clients/typescript/src/binary-transport.ts
import * as net from 'net';
import * as fs from 'fs';
import { BSON } from 'bson';

const HEADER_SIZE = 10;
const DEFAULT_SOCKET_PATH = '/tmp/mongocore.bin.sock';
const DEFAULT_MAX_FRAME_SIZE = 64 * 1024 * 1024;

// Opcodes
const OP_FIND_ONE = 0x02;
const OP_INSERT_ONE = 0x03;
const OP_INSERT_MANY = 0x04;
const OP_UPDATE_ONE = 0x05;
const OP_UPDATE_MANY = 0x06;
const OP_DELETE_ONE = 0x07;
const OP_DELETE_MANY = 0x08;
const OP_COUNT = 0x0A;
const OP_RUN_COMMAND = 0x0D;
const OP_HANDSHAKE = 0x3F;

// Flag bits
const FLAG_EOS = 1 << 6;
const FLAG_NO_REPLY = 1 << 7;

/**
 * Binary UDS transport client for MongoCore.
 *
 * Connects to MongoCore's binary protocol over Unix Domain Socket for
 * low-latency, high-throughput communication using BSON-encoded frames.
 */
export class BinaryTransport {
  private socketPath: string;
  private socket: net.Socket | null = null;
  private maxFrameSize: number = DEFAULT_MAX_FRAME_SIZE;
  private reqId: number = 0;
  private buffer: Buffer = Buffer.alloc(0);
  private dataResolve: ((value: void) => void) | null = null;

  constructor(socketPath?: string) {
    this.socketPath = socketPath
      || process.env.MONGOCORE_BINARY_SOCKET_PATH
      || DEFAULT_SOCKET_PATH;
  }

  /**
   * Connect to the MongoCore binary UDS socket and perform handshake.
   */
  async connect(): Promise<void> {
    await this.connectSocket();
    await this.handshake();
  }

  /**
   * Close the connection.
   */
  close(): void {
    if (this.socket) {
      this.socket.destroy();
      this.socket = null;
    }
    this.buffer = Buffer.alloc(0);
  }

  /**
   * Find a single document matching the filter.
   */
  async findOne(
    db: string,
    collection: string,
    filter: object,
    projection?: object,
  ): Promise<object | null> {
    const envelope: Record<string, unknown> = { db, coll: collection, filter };
    if (projection) {
      envelope.projection = projection;
    }

    const reqId = await this.sendFrame(OP_FIND_ONE, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    if (response.rawDocs.length === 0) {
      return null;
    }

    return BSON.deserialize(response.rawDocs) as object;
  }

  /**
   * Insert a single document. Returns the inserted ID as a string.
   */
  async insertOne(db: string, collection: string, document: object): Promise<string> {
    const envelope = { db, coll: collection };
    const rawDocs = Buffer.from(BSON.serialize(document));

    const reqId = await this.sendFrame(OP_INSERT_ONE, envelope, rawDocs);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    return result.inserted_id as string;
  }

  /**
   * Insert multiple documents. Returns the count of inserted documents.
   */
  async insertMany(
    db: string,
    collection: string,
    documents: object[],
    ordered?: boolean,
  ): Promise<number> {
    const envelope: Record<string, unknown> = { db, coll: collection };
    if (ordered !== undefined) {
      envelope.ordered = ordered;
    }

    // Concatenate BSON documents into raw docs buffer
    const parts = documents.map(doc => Buffer.from(BSON.serialize(doc)));
    const rawDocs = Buffer.concat(parts);

    const reqId = await this.sendFrame(OP_INSERT_MANY, envelope, rawDocs);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    return result.inserted_count as number;
  }

  /**
   * Delete a single document matching the filter. Returns the deleted count.
   */
  async deleteOne(db: string, collection: string, filter: object): Promise<number> {
    const envelope = { db, coll: collection, filter };

    const reqId = await this.sendFrame(OP_DELETE_ONE, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    return result.deleted_count as number;
  }

  /**
   * Count documents matching the filter.
   */
  async countDocuments(db: string, collection: string, filter?: object): Promise<number> {
    const envelope: Record<string, unknown> = { db, coll: collection };
    if (filter) {
      envelope.filter = filter;
    }

    const reqId = await this.sendFrame(OP_COUNT, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    return result.count as number;
  }

  /**
   * Update a single document matching the filter.
   * Returns matched and modified counts.
   */
  async updateOne(
    db: string,
    collection: string,
    filter: object,
    update: object,
  ): Promise<{ matchedCount: number; modifiedCount: number }> {
    const envelope = { db, coll: collection, filter, update, doc_bytes_len: 0 };

    await this.sendFrame(OP_UPDATE_ONE, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    return {
      matchedCount: result.matched_count as number,
      modifiedCount: result.modified_count as number,
    };
  }

  /**
   * Update multiple documents matching the filter.
   * Returns matched and modified counts.
   */
  async updateMany(
    db: string,
    collection: string,
    filter: object,
    update: object,
  ): Promise<{ matchedCount: number; modifiedCount: number }> {
    const envelope = { db, coll: collection, filter, update, doc_bytes_len: 0 };

    await this.sendFrame(OP_UPDATE_MANY, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    return {
      matchedCount: result.matched_count as number,
      modifiedCount: result.modified_count as number,
    };
  }

  /**
   * Delete multiple documents matching the filter. Returns the deleted count.
   */
  async deleteMany(db: string, collection: string, filter: object): Promise<number> {
    const envelope = { db, coll: collection, filter, doc_bytes_len: 0 };

    await this.sendFrame(OP_DELETE_MANY, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    return result.deleted_count as number;
  }

  /**
   * Run a raw database command. Returns the result document.
   */
  async runCommand(db: string, command: object): Promise<object> {
    const envelope = { db, command, doc_bytes_len: 0 };

    await this.sendFrame(OP_RUN_COMMAND, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(result.error as string);
    }

    if (response.rawDocs.length > 0) {
      return BSON.deserialize(response.rawDocs) as object;
    }

    return result as object;
  }

  /**
   * Check if the binary transport socket is available.
   */
  static isAvailable(socketPath?: string): boolean {
    const path = socketPath
      || process.env.MONGOCORE_BINARY_SOCKET_PATH
      || DEFAULT_SOCKET_PATH;
    try {
      fs.accessSync(path);
      return true;
    } catch {
      return false;
    }
  }

  // --- Private methods ---

  private async connectSocket(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.socket = net.createConnection(this.socketPath, () => {
        resolve();
      });

      this.socket.on('error', (err) => {
        reject(new Error(`Failed to connect to ${this.socketPath}: ${err.message}`));
      });

      this.socket.on('data', (chunk: Buffer) => {
        this.buffer = Buffer.concat([this.buffer, chunk]);
        if (this.dataResolve) {
          const resolve = this.dataResolve;
          this.dataResolve = null;
          resolve();
        }
      });
    });
  }

  private async handshake(): Promise<void> {
    const envelope = {
      client_language: 'typescript',
    };

    await this.sendFrame(OP_HANDSHAKE, envelope);
    const response = await this.readFrame();
    const result = BSON.deserialize(response.envelope);

    if (result.error) {
      throw new Error(`Handshake failed: ${result.error}`);
    }

    if (result.max_frame_size) {
      this.maxFrameSize = result.max_frame_size as number;
    }
  }

  private nextReqId(): number {
    this.reqId = (this.reqId + 1) & 0xFFFFFFFF;
    return this.reqId;
  }

  private async sendFrame(
    opcode: number,
    envelope: object,
    rawDocs?: Buffer,
    noReply?: boolean,
  ): Promise<number> {
    if (!this.socket) {
      throw new Error('Not connected');
    }

    const envelopeBuf = Buffer.from(BSON.serialize(envelope));
    const docsBuf = rawDocs || Buffer.alloc(0);
    const reqId = this.nextReqId();

    // msg_len = flags(2) + req_id(4) + envelope + rawDocs
    const msgLen = 2 + 4 + envelopeBuf.length + docsBuf.length;

    if (msgLen > this.maxFrameSize) {
      throw new Error(`Frame size ${msgLen} exceeds max frame size ${this.maxFrameSize}`);
    }

    // Build flags: bits[0:5] = opcode, bit6 = EOS, bit7 = no-reply
    let flags = (opcode & 0x3F) | FLAG_EOS;
    if (noReply) {
      flags |= FLAG_NO_REPLY;
    }

    // Header: 4B msg_len (BE) + 2B flags (BE) + 4B req_id (BE)
    const header = Buffer.alloc(HEADER_SIZE);
    header.writeUInt32BE(msgLen, 0);
    header.writeUInt16BE(flags, 4);
    header.writeUInt32BE(reqId, 6);

    const frame = Buffer.concat([header, envelopeBuf, docsBuf]);

    return new Promise((resolve, reject) => {
      this.socket!.write(frame, (err) => {
        if (err) return reject(err);
        resolve(reqId);
      });
    });
  }

  private async readFrame(): Promise<{ envelope: Buffer; rawDocs: Buffer }> {
    // Read header
    const header = await this.readExact(HEADER_SIZE);
    const msgLen = header.readUInt32BE(0);
    const flags = header.readUInt16BE(4);

    if (msgLen > this.maxFrameSize) {
      throw new Error(`Response frame size ${msgLen} exceeds max frame size ${this.maxFrameSize}`);
    }

    // Payload size = msgLen - flags(2) - req_id(4)
    const payloadSize = msgLen - 6;
    const payload = await this.readExact(payloadSize);

    // Check for error opcode (0x3E)
    const opcode = flags & 0x3F;
    if (opcode === 0x3E) {
      const envelopeSize = payload.readInt32LE(0);
      const errDoc = BSON.deserialize(payload.subarray(0, envelopeSize));
      throw new Error(errDoc.message || errDoc.error || 'Unknown server error');
    }

    // First 4 bytes of payload are LE i32 = BSON envelope size
    const envelopeSize = payload.readInt32LE(0);
    const envelope = payload.subarray(0, envelopeSize);
    const rawDocs = payload.subarray(envelopeSize);

    return { envelope, rawDocs };
  }

  private async readExact(n: number): Promise<Buffer> {
    while (this.buffer.length < n) {
      await this.waitForData();
    }

    const result = this.buffer.subarray(0, n);
    this.buffer = this.buffer.subarray(n);
    return result;
  }

  private waitForData(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.socket) {
        return reject(new Error('Socket closed'));
      }

      this.dataResolve = resolve;

      // Handle socket close/error while waiting
      const onClose = () => {
        this.dataResolve = null;
        reject(new Error('Socket closed while reading'));
      };
      const onError = (err: Error) => {
        this.dataResolve = null;
        reject(err);
      };

      this.socket.once('close', onClose);
      this.socket.once('error', onError);

      // Clean up listeners when resolved
      const originalResolve = this.dataResolve;
      this.dataResolve = () => {
        this.socket?.removeListener('close', onClose);
        this.socket?.removeListener('error', onError);
        resolve();
      };
    });
  }
}

/**
 * Connection pool for BinaryTransport with round-robin dispatch.
 *
 * Creates multiple connections to MongoCore's binary protocol and distributes
 * requests across them for improved throughput under concurrent workloads.
 */
export class BinaryTransportPool {
  private connections: BinaryTransport[] = [];
  private index = 0;
  private poolSize: number;
  private socketPath?: string;

  constructor(socketPath?: string, poolSize = 4) {
    this.socketPath = socketPath;
    this.poolSize = poolSize;
  }

  /**
   * Connect all pool members to the MongoCore binary UDS socket.
   */
  async connect(): Promise<void> {
    for (let i = 0; i < this.poolSize; i++) {
      const conn = new BinaryTransport(this.socketPath);
      await conn.connect();
      this.connections.push(conn);
    }
  }

  /**
   * Close all connections in the pool.
   */
  close(): void {
    for (const conn of this.connections) {
      conn.close();
    }
    this.connections = [];
  }

  private next(): BinaryTransport {
    const conn = this.connections[this.index % this.connections.length];
    this.index++;
    return conn;
  }

  // Proxy all methods via round-robin dispatch
  async findOne(...args: Parameters<BinaryTransport['findOne']>) { return this.next().findOne(...args); }
  async insertOne(...args: Parameters<BinaryTransport['insertOne']>) { return this.next().insertOne(...args); }
  async insertMany(...args: Parameters<BinaryTransport['insertMany']>) { return this.next().insertMany(...args); }
  async updateOne(...args: Parameters<BinaryTransport['updateOne']>) { return this.next().updateOne(...args); }
  async updateMany(...args: Parameters<BinaryTransport['updateMany']>) { return this.next().updateMany(...args); }
  async deleteOne(...args: Parameters<BinaryTransport['deleteOne']>) { return this.next().deleteOne(...args); }
  async deleteMany(...args: Parameters<BinaryTransport['deleteMany']>) { return this.next().deleteMany(...args); }
  async countDocuments(...args: Parameters<BinaryTransport['countDocuments']>) { return this.next().countDocuments(...args); }
  async runCommand(...args: Parameters<BinaryTransport['runCommand']>) { return this.next().runCommand(...args); }
}

File: clients/python/src/mongocore/binary_transport.py
"""Binary UDS transport for MongoCore — high-performance local protocol."""
import os
import struct
import socket
from typing import Optional, Any

import bson

HEADER_SIZE = 10
DEFAULT_SOCKET_PATH = "/tmp/mongocore.bin.sock"
DEFAULT_MAX_FRAME_SIZE = 64 * 1024 * 1024

# Opcodes (bits 0-5 of flags)
OP_FIND = 0x01
OP_FIND_ONE = 0x02
OP_INSERT_ONE = 0x03
OP_INSERT_MANY = 0x04
OP_UPDATE_ONE = 0x05
OP_UPDATE_MANY = 0x06
OP_DELETE_ONE = 0x07
OP_DELETE_MANY = 0x08
OP_AGGREGATE = 0x09
OP_COUNT = 0x0A
OP_RUN_COMMAND = 0x0D
OP_ERROR = 0x3E
OP_HANDSHAKE = 0x3F

# Flag bits (bit 0 = LSB)
FLAG_EOS = 1 << 6
FLAG_NO_REPLY = 1 << 7
FLAG_BATCH_FOLLOWS = 1 << 8


class BinaryTransportError(Exception):
    """Error from the binary transport layer."""
    pass


class BinaryTransport:
    """High-performance binary transport over Unix domain sockets.

    Connects to MongoCore's binary UDS listener for lower-latency,
    lower-overhead communication compared to gRPC.

    Usage:
        transport = BinaryTransport()
        transport.connect()
        doc = transport.find_one("mydb", "mycoll", {"name": "Alice"})
        transport.close()
    """

    def __init__(self, socket_path: Optional[str] = None):
        """Initialize the binary transport.

        Args:
            socket_path: Path to the Unix domain socket. Falls back to
                MONGOCORE_BINARY_SOCKET_PATH env var, then default path.
        """
        self._socket_path = (
            socket_path
            or os.environ.get("MONGOCORE_BINARY_SOCKET_PATH")
            or DEFAULT_SOCKET_PATH
        )
        self._sock: Optional[socket.socket] = None
        self._req_counter = 0

    def connect(self):
        """Connect to the MongoCore binary UDS and perform handshake.

        Raises:
            BinaryTransportError: If connection or handshake fails.
        """
        try:
            self._sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            self._sock.connect(self._socket_path)
        except (OSError, ConnectionRefusedError) as e:
            self._sock = None
            raise BinaryTransportError(f"Failed to connect to {self._socket_path}: {e}") from e

        self._handshake()
        return self

    def close(self):
        """Close the socket connection."""
        if self._sock:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None

    def find(
        self,
        db: str,
        collection: str,
        filter: Optional[dict] = None,
        limit: int = 0,
    ) -> list[dict]:
        """Find multiple documents matching the filter.

        Args:
            db: Database name.
            collection: Collection name.
            filter: Query filter document (matches all if None/empty).
            limit: Maximum number of documents to return (0 = no limit).

        Returns:
            List of matched documents.
        """
        envelope: dict[str, Any] = {"db": db, "coll": collection, "filter": filter or {}}
        if limit:
            envelope["limit"] = limit

        self._send_frame(OP_FIND, envelope)
        resp_envelope, raw_docs = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])

        # Parse concatenated BSON docs from raw_docs
        docs = []
        offset = 0
        while offset < len(raw_docs):
            doc_len = struct.unpack("<i", raw_docs[offset:offset + 4])[0]
            docs.append(bson.decode(raw_docs[offset:offset + doc_len]))
            offset += doc_len
        return docs

    def find_one(
        self,
        db: str,
        collection: str,
        filter: dict,
        projection: Optional[dict] = None,
    ) -> Optional[dict]:
        """Find a single document matching the filter.

        Args:
            db: Database name.
            collection: Collection name.
            filter: Query filter document.
            projection: Optional projection document.

        Returns:
            The matched document, or None if no match.
        """
        envelope = {"db": db, "coll": collection, "filter": filter}
        if projection:
            envelope["projection"] = projection

        self._send_frame(OP_FIND_ONE, envelope)
        resp_envelope, raw_docs = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])

        if not raw_docs:
            return None
        return bson.decode(raw_docs)

    def insert_one(self, db: str, collection: str, document: dict) -> str:
        """Insert a single document.

        Args:
            db: Database name.
            collection: Collection name.
            document: The document to insert.

        Returns:
            The inserted document's _id as a string.
        """
        raw_doc = bson.encode(document)
        envelope = {"db": db, "coll": collection}

        self._send_frame(OP_INSERT_ONE, envelope, raw_docs=raw_doc)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])
        return str(resp.get("inserted_id", ""))

    def insert_many(
        self,
        db: str,
        collection: str,
        documents: list[dict],
        ordered: bool = True,
    ) -> int:
        """Insert multiple documents.

        Args:
            db: Database name.
            collection: Collection name.
            documents: List of documents to insert.
            ordered: Whether to stop on first error.

        Returns:
            The number of documents inserted.
        """
        raw_docs = b"".join(bson.encode(doc) for doc in documents)
        envelope = {"db": db, "coll": collection, "ordered": ordered, "count": len(documents)}

        self._send_frame(OP_INSERT_MANY, envelope, raw_docs=raw_docs)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])
        return resp.get("inserted_count", 0)

    def update_one(self, db: str, collection: str, filter: dict, update: dict) -> dict:
        """Update a single document.

        Args:
            db: Database name.
            collection: Collection name.
            filter: Query filter document.
            update: Update operations document.

        Returns:
            Dict with matched_count and modified_count.
        """
        envelope = {"db": db, "coll": collection, "filter": filter, "update": update}

        self._send_frame(OP_UPDATE_ONE, envelope)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])
        return {
            "matched_count": resp.get("matched_count", 0),
            "modified_count": resp.get("modified_count", 0),
        }

    def update_many(self, db: str, collection: str, filter: dict, update: dict) -> dict:
        """Update multiple documents.

        Args:
            db: Database name.
            collection: Collection name.
            filter: Query filter document.
            update: Update operations document.

        Returns:
            Dict with matched_count and modified_count.
        """
        envelope = {"db": db, "coll": collection, "filter": filter, "update": update}

        self._send_frame(OP_UPDATE_MANY, envelope)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])
        return {
            "matched_count": resp.get("matched_count", 0),
            "modified_count": resp.get("modified_count", 0),
        }

    def delete_one(self, db: str, collection: str, filter: dict) -> int:
        """Delete a single document.

        Args:
            db: Database name.
            collection: Collection name.
            filter: Query filter document.

        Returns:
            Number of documents deleted (0 or 1).
        """
        envelope = {"db": db, "coll": collection, "filter": filter}

        self._send_frame(OP_DELETE_ONE, envelope)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])
        return resp.get("deleted_count", 0)

    def delete_many(self, db: str, collection: str, filter: dict) -> int:
        """Delete multiple documents.

        Args:
            db: Database name.
            collection: Collection name.
            filter: Query filter document.

        Returns:
            Number of documents deleted.
        """
        envelope = {"db": db, "coll": collection, "filter": filter}

        self._send_frame(OP_DELETE_MANY, envelope)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])
        return resp.get("deleted_count", 0)

    def count_documents(self, db: str, collection: str, filter: Optional[dict] = None) -> int:
        """Count documents matching a filter.

        Args:
            db: Database name.
            collection: Collection name.
            filter: Optional query filter (counts all if None).

        Returns:
            The document count.
        """
        envelope = {"db": db, "coll": collection, "filter": filter or {}}

        self._send_frame(OP_COUNT, envelope)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])
        return resp.get("count", 0)

    def run_command(self, db: str, command: dict) -> dict:
        """Run an arbitrary database command.

        Args:
            db: Database name.
            command: The command document.

        Returns:
            The command response document.
        """
        envelope = {"db": db, "command": command}

        self._send_frame(OP_RUN_COMMAND, envelope)
        resp_envelope, raw_docs = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(resp["error"])

        if raw_docs:
            return bson.decode(raw_docs)
        return resp

    @staticmethod
    def is_available(socket_path: Optional[str] = None) -> bool:
        """Check if the binary transport socket exists and is connectable.

        Args:
            socket_path: Path to check. Falls back to env var, then default.

        Returns:
            True if the socket exists and accepts connections.
        """
        path = (
            socket_path
            or os.environ.get("MONGOCORE_BINARY_SOCKET_PATH")
            or DEFAULT_SOCKET_PATH
        )
        if not os.path.exists(path):
            return False

        try:
            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            sock.settimeout(1.0)
            sock.connect(path)
            sock.close()
            return True
        except (OSError, ConnectionRefusedError):
            return False

    # --- Internal methods ---

    def _handshake(self):
        """Perform protocol handshake with the server."""
        envelope = {"client_language": "python"}
        self._send_frame(OP_HANDSHAKE, envelope)
        resp_envelope, _ = self._read_frame()

        resp = bson.decode(resp_envelope)
        if resp.get("error"):
            raise BinaryTransportError(f"Handshake failed: {resp['error']}")

    def _next_req_id(self) -> int:
        """Generate the next monotonic request ID."""
        self._req_counter += 1
        return self._req_counter

    def _send_frame(
        self,
        opcode: int,
        envelope_dict: dict,
        raw_docs: bytes = b"",
        no_reply: bool = False,
    ) -> int:
        """Send a framed message to the server.

        Args:
            opcode: Operation code (bits 0-5 of flags).
            envelope_dict: Envelope to BSON-encode.
            raw_docs: Raw BSON document bytes to append after envelope.
            no_reply: If True, set the no-reply flag.

        Returns:
            The request ID used for this frame.
        """
        if self._sock is None:
            raise BinaryTransportError("Not connected. Call connect() first.")

        req_id = self._next_req_id()
        envelope_bytes = bson.encode(envelope_dict)

        # flags: bits[0:5] = opcode, bit6 = EOS (always set for single frame),
        #         bit7 = no-reply, bit8 = batch-follows
        flags = (opcode & 0x3F) | FLAG_EOS
        if no_reply:
            flags |= FLAG_NO_REPLY

        # msg_len = flags(2) + req_id(4) + envelope + raw_docs
        msg_len = 2 + 4 + len(envelope_bytes) + len(raw_docs)

        # Header: 4B msg_len (BE) + 2B flags (BE) + 4B req_id (BE)
        header = struct.pack(">IHI", msg_len, flags, req_id)
        self._sock.sendall(header + envelope_bytes + raw_docs)

        return req_id

    def _read_frame(self) -> tuple[bytes, bytes]:
        """Read a complete frame from the server.

        Returns:
            Tuple of (envelope_bytes, raw_doc_bytes).

        Raises:
            BinaryTransportError: On protocol errors or disconnection.
        """
        # Read header
        header = self._recv_exact(HEADER_SIZE)
        msg_len, flags, req_id = struct.unpack(">IHI", header)

        # Validate
        payload_len = msg_len - 2 - 4  # subtract flags + req_id sizes
        if payload_len < 0:
            raise BinaryTransportError(f"Invalid frame: msg_len={msg_len} too small")
        if payload_len > DEFAULT_MAX_FRAME_SIZE:
            raise BinaryTransportError(f"Frame too large: {payload_len} bytes")

        opcode = flags & 0x3F
        if opcode == OP_ERROR:
            # Read error payload
            payload = self._recv_exact(payload_len) if payload_len > 0 else b""
            if payload:
                bson_len = struct.unpack("<i", payload[:4])[0]
                err = bson.decode(payload[:bson_len])
                raise BinaryTransportError(err.get("message", err.get("error", "Unknown server error")))
            raise BinaryTransportError("Unknown server error")

        # Read payload
        if payload_len == 0:
            return b"", b""

        payload = self._recv_exact(payload_len)

        # Split envelope from raw docs using BSON document length prefix
        if len(payload) < 4:
            return payload, b""

        # First 4 bytes of BSON doc is its length (little-endian i32)
        envelope_len = struct.unpack("<i", payload[:4])[0]
        if envelope_len < 5 or envelope_len > len(payload):
            raise BinaryTransportError(
                f"Invalid envelope length: {envelope_len} (payload={len(payload)})"
            )

        envelope_bytes = payload[:envelope_len]
        raw_doc_bytes = payload[envelope_len:]
        return envelope_bytes, raw_doc_bytes

    def _recv_exact(self, n: int) -> bytes:
        """Read exactly n bytes from the socket.

        Args:
            n: Number of bytes to read.

        Returns:
            Exactly n bytes.

        Raises:
            BinaryTransportError: If connection closed before all bytes read.
        """
        if self._sock is None:
            raise BinaryTransportError("Not connected.")

        chunks = []
        remaining = n
        while remaining > 0:
            chunk = self._sock.recv(min(remaining, 65536))
            if not chunk:
                raise BinaryTransportError(
                    f"Connection closed (read {n - remaining}/{n} bytes)"
                )
            chunks.append(chunk)
            remaining -= len(chunk)
        return b"".join(chunks)

    def __enter__(self):
        self.connect()
        return self

    def __exit__(self, *args):
        self.close()


class BinaryTransportPool:
    """Connection pool for binary transport — distributes operations across multiple connections.

    Maintains a pool of BinaryTransport connections and distributes requests
    using round-robin selection. Default pool size of 4 connections supports
    high-throughput CPU-bound workloads.

    Usage:
        with BinaryTransportPool(pool_size=4) as pool:
            doc = pool.find_one("mydb", "mycoll", {"name": "Alice"})
    """

    def __init__(self, socket_path: Optional[str] = None, pool_size: int = 4):
        """Initialize the connection pool.

        Args:
            socket_path: Path to the Unix domain socket. Falls back to
                MONGOCORE_BINARY_SOCKET_PATH env var, then default path.
            pool_size: Number of connections to maintain (default: 4).
        """
        self._socket_path = (
            socket_path
            or os.environ.get("MONGOCORE_BINARY_SOCKET_PATH")
            or DEFAULT_SOCKET_PATH
        )
        self._pool_size = pool_size
        self._connections: list[BinaryTransport] = []
        self._index = 0

    def connect(self):
        """Open all connections in the pool.

        Raises:
            BinaryTransportError: If any connection fails (all opened
                connections are closed on failure).
        """
        try:
            for _ in range(self._pool_size):
                conn = BinaryTransport(self._socket_path)
                conn.connect()
                self._connections.append(conn)
        except BinaryTransportError:
            self.close()
            raise
        return self

    def close(self):
        """Close all connections in the pool."""
        for conn in self._connections:
            conn.close()
        self._connections.clear()
        self._index = 0

    def _next(self) -> BinaryTransport:
        """Round-robin connection selection."""
        if not self._connections:
            raise BinaryTransportError("Pool not connected. Call connect() first.")
        conn = self._connections[self._index % len(self._connections)]
        self._index += 1
        return conn

    def find(
        self,
        db: str,
        collection: str,
        filter: Optional[dict] = None,
        limit: int = 0,
    ) -> list[dict]:
        """Find multiple documents matching the filter."""
        return self._next().find(db, collection, filter, limit)

    def find_one(
        self,
        db: str,
        collection: str,
        filter: dict,
        projection: Optional[dict] = None,
    ) -> Optional[dict]:
        """Find a single document matching the filter."""
        return self._next().find_one(db, collection, filter, projection)

    def insert_one(self, db: str, collection: str, document: dict) -> str:
        """Insert a single document."""
        return self._next().insert_one(db, collection, document)

    def insert_many(
        self,
        db: str,
        collection: str,
        documents: list[dict],
        ordered: bool = True,
    ) -> int:
        """Insert multiple documents."""
        return self._next().insert_many(db, collection, documents, ordered)

    def update_one(self, db: str, collection: str, filter: dict, update: dict) -> dict:
        """Update a single document."""
        return self._next().update_one(db, collection, filter, update)

    def update_many(self, db: str, collection: str, filter: dict, update: dict) -> dict:
        """Update multiple documents."""
        return self._next().update_many(db, collection, filter, update)

    def delete_one(self, db: str, collection: str, filter: dict) -> int:
        """Delete a single document."""
        return self._next().delete_one(db, collection, filter)

    def delete_many(self, db: str, collection: str, filter: dict) -> int:
        """Delete multiple documents."""
        return self._next().delete_many(db, collection, filter)

    def count_documents(self, db: str, collection: str, filter: Optional[dict] = None) -> int:
        """Count documents matching a filter."""
        return self._next().count_documents(db, collection, filter)

    def run_command(self, db: str, command: dict) -> dict:
        """Run an arbitrary database command."""
        return self._next().run_command(db, command)

    def __enter__(self):
        self.connect()
        return self

    def __exit__(self, *args):
        self.close()
